### ISSUE TITLE
from_iter_instead_of_collect / needless_collect interaction questionable

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6533]
````rust
for _i in Vec::from_iter([1, 2, 3].iter()) {}
````
clippy will suggest to use `.collect()` instead of `from_iter()` here but blindly replacing will not work because `.collect()` will need some kind of type hint.

Them again
````rust
for _i in [1, 2, 3].iter().collect::<Vec<&i32>>() {}
````
does not trigger `clippy::needless_collect` (although it should!) because we can iterate over
````rust
for _i in [1, 2, 3].iter() {}
````
directly.

**Comment by camsteffen** [comment]
Re-labeling as the first issue is fixed in #6657.

**Comment by y21** [comment]
Seems to be fixed, the second case also lints now: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=99d5028b5d63461da7aeb8466999d9d4
